### DISCUSSION TITLE
How to insert API keys into frontend bundle

### DISCUSSION THREAD

**Discussion by theryan722** [parcel-bundler/parcel#5915] [Q&A] [answered]
Hello, I have a TypeScript project that I am using with Parcel. The old way I used to insert public facing API keys (e.g. for Firebase) in the client-side bundle was using Gulp to modify the file and replace strings in the files at compilation with the api keys from the config file.

How would I do this with Parcel?

I am running parcel through a Gulp task using the Parcel API.

It seems expensive to wait for parcel to build the entire bundle(s), then open up each file and search for the strings to replace.

Thanks!

**Comment by mischnic** [comment, marked as answer]
You can use `process.env` in your frontend code and Parcel will replace it:

```js
console.log(process.env.KEYS)
```
And then set that variable when invoking Parcel
`KEYS=something parcel build index.js`

**Reply by theryan722** [reply]
Thank you for the fast and helpful reply!

**Reply by qqqq-collab** [reply]
ok

**Comment by FDiskas** [comment]
What about `.env` files?

Parcel v1: https://parceljs.org/env.html

parcel v2
https://v2.parceljs.org/features/node-emulation/#%F0%9F%8C%B3-environment-variables
